monitor/monitor.go

```diff
@@ -44,54 +45,55 @@ type Monitor struct {
 
 // Run starts monitoring.
 func (m *Monitor) Run(npages int) {
-	log.Info("Starting monitoring traffic")
-	c := bpf.DefaultPerfEventConfig()
-	c.NumPages = npages
-
-	me, err := bpf.NewPerCpuEvents(c)
-	if err != nil {
-		log.WithError(err).Error("Error while starting monitor")
-		return
-	}
-	monitorEvents = me
-
-	// Dump stat
+	eventConfig.NumPages = npages
 	t := time.NewTicker(5 * time.Second)
-	go func(t *time.Ticker) {
-		for {
-			select {
-			case <-t.C:
-				if monitorEvents != nil {
-					m.dumpStat()
-				}
-			}
+	for range t.C {
+		// This loop is meant to loop forever so our process does not become
+		// defunct and we still print the stat for connected monitors.
+		if monitorEvents != nil {
+			m.dumpStat()
 		}
-	}(t)
+	}
+}
 
-	// Main event loop
-	for {
-		todo, err := monitorEvents.Poll(pollTimeout)
-		if err != nil {
-			log.WithError(err).Error("Error in Poll")
-			if err == syscall.EBADF {
-				break
-			}
-		}
-		if todo > 0 {
-			if err := monitorEvents.ReadAll(m.receiveEvent, m.lostEvent); err != nil {
-				log.WithError(err).Warn("Error received while reading from perf buffer")
-			}
+func purgeInactiveListeners() {
+	metaBuf := generateHealthCheckMsg()
+	var next *list.Element
+	for e := listeners.Front(); e != nil; e = next {
+		client := e.Value.(net.Conn)
+		next = e.Next()
+		if _, err := client.Write(metaBuf); err != nil {
+			log.WithError(err).Warn("metadata write failed; removing client")
+			client.Close()
+			listeners.Remove(e)
+			continue
 		}
 	}
 }
 
+func generateHealthCheckMsg() []byte {
```

Comment:
Oh, it just occurred to me that since we just keep sending this same empty meta, and only in `purgeInactiveListeners`, we can store the `[]byte` and just keep reusing that. It's not a big deal, though, this isn't a big memory use or anything.